src: always compile and store code cache for native modules

joyeecheung opened the pull request:
#### src: remove code cache integrity check

In preparation of sharing code cache among different threads -
we simply rely on v8 to reject invalid cache, since there isn't
any serious consequence when the cache is invalid anyway.

#### src: always compile and store code cache for native modules

This patch changes the NativeModuleLoader to always try to find
code cache for native modules when it compiles them, and always
produce and store the code cache after compilation. The cache
map is protected by a mutex and can be accessed by different
threads - including the worker threads and the main thread. Hence any
thread can reuse the code cache if the native module has already
been compiled by another thread - in particular the cache of the
bootstrappers and per_context.js will always be hit when a new thread
is spun.

This results in a ~6% startup overhead in the worst case
(when only the main thread is launched without requiring any additional
native module - it now needs to do the extra work of finding and
storing caches), which balances out the recent improvements by moving
the compilation to C++, but it also leads to a ~60% improvement in
the best case (when a worker thread is spun and requires a lot of native
modules thus hitting the cache compiled by the main thread).

Local benchmark results:

```
                                                                                   confidence improvement accuracy (*)   (**)  (***)
 misc/startup.js mode='process' script='benchmark/fixtures/require-cachable' dur=1        ***     -6.80 %       ±0.77% ±1.02% ±1.34%
 misc/startup.js mode='process' script='test/fixtures/semicolon' dur=1                    ***     -4.90 %       ±0.82% ±1.10% ±1.43%
 misc/startup.js mode='worker' script='benchmark/fixtures/require-cachable' dur=1         ***     63.13 %       ±1.58% ±2.12% ±2.79%
 misc/startup.js mode='worker' script='test/fixtures/semicolon' dur=1                     ***     56.20 %       ±1.33% ±1.78% ±2.33%
```

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

joyeecheung commented:
CI: https://ci.nodejs.org/job/node-test-pull-request/19408/
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/281/

addaleax commented on src/node_native_module.cc:
```diff
@@ -96,31 +102,54 @@ Local<String> NativeModuleLoader::GetConfigString(Isolate* isolate) const {
-void NativeModuleLoader::CompileCodeCache(
-    const FunctionCallbackInfo<Value>& args) {
+// This is supposed to be run only by the main thread in
+// tools/generate_code_cache.js
```
Should we add `CHECK(env->is_main_thread());` to enforce this?

addaleax commented on src/node_native_module.cc:
```diff
@@ -96,31 +102,54 @@ Local<String> NativeModuleLoader::GetConfigString(Isolate* isolate) const {
+
+  // The module has not been compiled before.
+  if (cached_data == nullptr) {
+    return MaybeLocal<Uint8Array>();
```
You could also `return` in the previous conditional if you flip to `it == code_cache_.end()`, right?

addaleax commented on src/node_native_module.cc:
```diff
@@ -244,10 +251,14 @@ MaybeLocal<Value> NativeModuleLoader::LookupAndCompile(
   Local<Function> fun = maybe_fun.ToLocalChecked();
+  // XXX(joyeecheung): this bookkeeping is not exactly accurate because
+  // it only starts after the Environment is created, so the per_context.js
+  // will never be any of these two sets, but the two sets are only for
```
Is this a typo (`never be in any of these sets`)?

joyeecheung commented:
@addaleax Thanks for the review, updated. Also updated the `emplace()` calls since gcc 4.9.4 needs the pointers to be moved before the call, and updated the test to pass the `--worker` test suit since `compileWithCache` is no longer guaranteed to be empty when a worker thread is spun (that's the whole point of this patch :p)

PTAL.

Also cc @nodejs/process @nodejs/build-files 

CI: https://ci.nodejs.org/job/node-test-pull-request/19412/
Benchmark CI: https://ci.nodejs.org/view/Node.js%20benchmark/job/benchmark-node-micro-benchmarks/282/

joyeecheung commented:
Removing `author-ready` because this needs at least one more approval to land.

From the previous benchmark CI:

```
                                                                                   confidence improvement accuracy (*)   (**)  (***)
 misc/startup.js mode='process' script='benchmark/fixtures/require-cachable' dur=1        ***     -7.38 %       ±1.09% ±1.46% ±1.90%
 misc/startup.js mode='process' script='test/fixtures/semicolon' dur=1                    ***     -5.12 %       ±0.79% ±1.06% ±1.37%
 misc/startup.js mode='worker' script='benchmark/fixtures/require-cachable' dur=1         ***     54.77 %       ±1.42% ±1.91% ±2.50%
 misc/startup.js mode='worker' script='test/fixtures/semicolon' dur=1                     ***     48.22 %       ±3.27% ±4.40% ±5.82%
```

Resume CI: https://ci.nodejs.org/job/node-test-pull-request/19454/

BridgeAR commented:
@joyeecheung the label is not meant as `ready to land` but `the author is done`. Missing LGs are absolutely fine while having the `author ready` label applied :-)

joyeecheung commented:
@BridgeAR 

> the label is not meant as ready to land but the author is done. Missing LGs are absolutely fine while having the author ready label applied :-)

That seems to contradict the label description:

> PRs that have at least two approvals, no pending requests for changes, and a CI started.

And the collaborator guide:

> at least two Collaborators approved the PR (one Collaborator approval is enough if the pull request has been open for more than 7 days).

BridgeAR commented:
@joyeecheung 
> That seems to contradict the label description:

The label was not up to date. It's fixed now.

> And the collaborator guide:

Thanks! Seems like I missed that in #24893. I am opening a PR to fix that right away.

joyeecheung commented:
CI before landing: https://ci.nodejs.org/job/node-test-pull-request/19636/

joyeecheung commented:
Landed in ceb66352236240498a9263500c7bde74e58c71fa and 0858e5d9d8db085cb83b3f1f3f94ed6b550a7bc5

MylesBorins commented:
This does not land cleanly on v11.x, would someone be willing to backport?

BridgeAR commented:
Ping @joyeecheung

addaleax commented:
Lands cleanly now :+1:
